### PR TITLE
DEL-272 monitor 3scale

### PR DESCRIPTION
**Pull request by b1zzu**
## Verify

1. Build the workload-app image and push it to a public repo 
   ```
   docker build -t quay.io/myrepo/workload-app .
   docker push quay.io/myrepo/workload-app
   ```
   or use mines `quay.io/b1zzu/workload-app`
2. Change the image in the `deploy/template.yaml` file
3. Login to OpenShift (ping me for the cluster)
4. Run `make local/deploy`
5. Check the logs of the deployed workload-app (it shouldn't print any failed requests messages)
6. Check also the /metrics endpoint (expose it using a route or a `oc proxy`)
7. Try to generate some error by changing the User Key of the workload-app-api in 3scale

### CODE REVIEW

**Comment by STEPHINRACHEL** [comment]
I tried to follow the steps to verify and after step  4 got timeout error and THREE_SCALE_URL seems to be empty
```Deploying the webapp with the following parameters:
AMQ_ADDRESS=amqps://messaging-eecbf30.redhat-rhmi-amq-online.svc
AMQ_QUEUE=/queue-requests
THREE_SCALE_URL=
deploymentconfig.apps.openshift.io/workload-web-app configured
service/workload-web-app unchanged
Waiting for pod to be ready
error: timed out waiting for the condition
make: *** [deploy] Error 1```

**Comment by b1zzu** [comment]
@STEPHINRACHEL do you have more logs? If something went wrong it should have print some error  before

**Comment by STEPHINRACHEL** [comment]
Screenshot 2020-05-27 at 18 31 47

**Comment by b1zzu** [comment]
I see the `base64` cmd is different in Mac, I would try to fix it and make it compatible on both system, can you also check if you have the `xpath` cmd? or the `xmllint` cmd?

**Comment by STEPHINRACHEL** [comment]
@b1zzu Yes I do have both `xpath` and `xmllint` cmds

**Comment by wei-lee** [comment]
@b1zzu I am see errors around xpath too:

```
Creating required AMQ resources
authenticationservice.admin.enmasse.io/none-authservice unchanged
addressspace.enmasse.io/workload-app unchanged
address.enmasse.io/workload-app.queue-requests unchanged
Waiting for AMQ AddressSpace to be ready
true
Waiting for AMQ Address to be ready
true
3scale.sh: API_URL=https://3scale-admin.apps.cluster-rhmi-3626.rhmi-3626.example.opentlc.com/admin/api
3scale.sh: create workload-app-api-account
Cannot open file '-q' at /System/Library/Perl/Extras/5.18/XML/XPath.pm line 53.
(23) Failed writing body
3scale.sh: create workload-app-api-backend
3scale.sh: create workload-app-api-metric in backend=3
3scale.sh: create mapping_rule with backend_id=3 and metric_id=9
3scale.sh: create workload-app-api service
Cannot open file '-q' at /System/Library/Perl/Extras/5.18/XML/XPath.pm line 53.
(23) Failed writing body
3scale.sh: create backend_usage with service_id= and backend_id=3
curl: (22) The requested URL returned error: 404 Not Found
Deploying the webapp with the following parameters:
AMQ_ADDRESS=amqps://messaging-iiq5def4ng.redhat-rhmi-amq-online.svc
AMQ_QUEUE=/queue-requests
THREE_SCALE_URL=
deploymentconfig.apps.openshift.io/workload-web-app configured
service/workload-web-app unchanged
Waiting for pod to be ready
pod/workload-web-app-5-kdd7d condition met
```

I think maybe we can build a container image and run the script from there?

**Comment by b1zzu** [comment]
@wei-lee can you send me an `xpath --help`, I would like to see if also xpath is different from linux to mac

**Comment by wei-lee** [comment]
@b1zzu Hmm, for some reason when I try `xpath --help` on mac, it just got stuck and outputs nothing..

**Comment by b1zzu** [comment]
@wei-lee I think it was an issue with the `| xpath` where `xpath` was closing the pipe before `curl` could write to it, I did some changes that should fix the problem and also give some more comprehensible errors if it fails again, can you retry on mac?

**Comment by b1zzu** [comment]
@STEPHINRACHEL you can also try again, the base64 error should be fix

**Comment by wei-lee** [comment]
@b1zzu tried it again and see the following output:

```
❯ make deploy
./deploy/deploy.sh
Error from server (AlreadyExists): project.project.openshift.io "workload-web-app" already exists
Creating required AMQ resources
authenticationservice.admin.enmasse.io/none-authservice unchanged
addressspace.enmasse.io/workload-app unchanged
address.enmasse.io/workload-app.queue-requests unchanged
Waiting for AMQ AddressSpace to be ready
true
Waiting for AMQ Address to be ready
true
3scale.sh: API_URL=https://3scale-admin.apps.cluster-rhmi-3626.rhmi-3626.example.opentlc.com/admin/api
3scale.sh: create workload-app-api-account
Cannot open file '-q' at /System/Library/Perl/Extras/5.18/XML/XPath.pm line 53.
3scale.sh: error: xget: failed to find '//account/id/text()' in: '<?xml version="1.0" encoding="UTF-8"?><errors><error>Users invalid</error><error>Username has already been taken</error></errors>'
3scale.sh: create workload-app-api-backend
3scale.sh: create workload-app-api-metric in backend=null
3scale.sh: create mapping_rule with backend_id=null and metric_id=null
3scale.sh: create workload-app-api service
Cannot open file '-q' at /System/Library/Perl/Extras/5.18/XML/XPath.pm line 53.
3scale.sh: error: xget: failed to find '//service/id/text()' in: '<?xml version="1.0" encoding="UTF-8"?><errors><error>System name has already been taken</error></errors>'
3scale.sh: create backend_usage with service_id= and backend_id=null
3scale.sh: create workload-app-api-plan in service=
Cannot open file '-q' at /System/Library/Perl/Extras/5.18/XML/XPath.pm line 53.
3scale.sh: error: xget: failed to find '//plan/id/text()' in: '<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>Not Found</title>
    <link href="/assets/error.css" media="screen" rel="stylesheet" type="text/css" />
  </head>

  <body>
    <div id="wrapper">
      <div id="content">
        <h1 class='logo-3scale logo-3scale--svg logo-3scale--boxed' id="logo">
          3scale
        </h1>

        <h1>Not Found</h1>
        <p>Sorry. We can't find what you're looking for.</p>
      </div>

      <!-- AIRBRAKE ERROR -->

      <div id="footer">
        <ul>
          <li class="powered-by-3scale last">
            <a href="http://www.3scale.net" target="_blank">
              Powered by <span class="logo-3scale logo-3scale--svg logo-3scale--powered">3scale</span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </body>
</html>'
3scale.sh: create workload-app-api-app in account= with plan_id=
Cannot open file '-q' at /System/Library/Perl/Extras/5.18/XML/XPath.pm line 53.
3scale.sh: error: xget: failed to find '//application/user_key/text()' in: '<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>Not Found</title>
    <link href="/assets/error.css" media="screen" rel="stylesheet" type="text/css" />
  </head>

  <body>
    <div id="wrapper">
      <div id="content">
        <h1 class='logo-3scale logo-3scale--svg logo-3scale--boxed' id="logo">
          3scale
        </h1>

        <h1>Not Found</h1>
        <p>Sorry. We can't find what you're looking for.</p>
      </div>

      <!-- AIRBRAKE ERROR -->

      <div id="footer">
        <ul>
          <li class="powered-by-3scale last">
            <a href="http://www.3scale.net" target="_blank">
              Powered by <span class="logo-3scale logo-3scale--svg logo-3scale--powered">3scale</span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </body>
</html>'
Deploying the webapp with the following parameters:
AMQ_ADDRESS=amqps://messaging-iiq5def4ng.redhat-rhmi-amq-online.svc
AMQ_QUEUE=/queue-requests
THREE_SCALE_URL=
deploymentconfig.apps.openshift.io/workload-web-app configured
service/workload-web-app unchanged
Waiting for pod to be ready
pod/workload-web-app-5-kdd7d condition met
pod/workload-web-app-6-gtpdd condition met
```

**Comment by wei-lee** [comment]
@b1zzu actually it could be a problem with the cluster I am running against. Do you have a cluster that I can try?

**Comment by STEPHINRACHEL** [comment]
@b1zzu 
I also got the same error as Wei's
https://gist.github.com/STEPHINRACHEL/ce9dfb2109e6390e1c073839d450a29f

**Comment by b1zzu** [comment]
@wei-lee I see two errors: 
```
Cannot open file '-q' at /System/Library/Perl/Extras/5.18/XML/XPath.pm line 53.
```
and 
```
<?xml version="1.0" encoding="UTF-8"?><errors><error>Users invalid</error><error>Username has already been taken</error></errors>
```
for the first, I don't know how to fix it because is a Mac OS thing, it seems like xpath for mac doesn't read from stdin or that it needs a special flag to read from stdin or that the `-q` option is not supported on mac, but I can't figure it out because I don't have a mac :(

the second error is because the cluster is dirty, just try `./deploy/3scale.sh undeploy`, but after you may need to wait ~10m before 3scale clean all resources

if you need a cluster you can use this one (the credentials are inside the job output): 
https://master-jenkins-csb-intly.cloud.paas.psi.redhat.com/job/Public/job/rhmi-install-latest/130/console

I should have implemented a search or create a system, but I wanted to avoid writing complex and repetitive logics in bash. Maybe I should rewrite it in golang, wdyt?

**Comment by wei-lee** [comment]
@b1zzu why not just build a container image? We can either update the existing one, or build a new one. I hope we will be able to use CRs to setup 3scale soon.

**Comment by mikenairn** [comment]
I have been unable to get this running, hit errors with xpath initially and then hit other issues:

```
3scale.sh: error: xget: failed to find '//plan/id/text()' in: '<!DOCTYPE html>
```

@wei-lee mentions above using CRs to create resources, is that something we could do instead? Use the 3scale operator itself to create the desired resources?

**Comment by b1zzu** [comment]
@mikenairn the 3scale CRs covers only half of the resource we need to create (they can't create an account and application), plus they are experimental, they will be refactored in the next version and they don't report any error when they fail and they failed a lot when I was trying them, so I thought to just use `curl`. It would make sense to start using CRs when they will become stable.

Can you pass me the full log, maybe it failed also before

**Comment by b1zzu** [comment]
@wei-lee @mikenairn I've created the tools container. Is now possible to run the deployment in the tools container locally using `make local/deploy`

File: Makefile [hunk 1]
```diff
@@ -1,10 +1,29 @@
 BUILD_TARGET=workload-app
 NAMESPACE=workload-web-app
+CONTAINER_ENGINE=docker
+TOOLS_IMAGE=quay.io/integreatly/workload-web-app-tools
```
Comment (mikenairn) [inline Makefile:4]: Has this image been built and pushed anywhere?
I guess we need to add this an extra image to build on CI also?

File: Dockerfile.tools [hunk 2]
```diff
@@ -0,0 +1,15 @@
+FROM centos:7
+
+COPY resources/rhit-root-ca.crt /etc/pki/ca-trust/source/anchors/
+RUN update-ca-trust extract
```
Comment (mikenairn) [inline Dockerfile.tools:4]: Was this needed?

**Comment by mikenairn** [comment]
@b1zzu Getting the following:

```
docker run --rm -it -e KUBECONFIG=/kube.config -v "/home/mnairn/.kube/config":/kube.config:z -v "/home/mnairn/go/src/github.com/integr8ly/workload-web-app":/workload-web-app -w /workload-web-app quay.io/integreatly/workload-web-app-tools make deploy
./deploy/deploy.sh
Already on project "workload-web-app" on server "https://api.mnairn.l9z7.s1.devshift.org:6443".

You can add applications to this project with the 'new-app' command. For example, try:

    oc new-app ruby~https://github.com/sclorg/ruby-ex.git

to build a new example application in Ruby. Or use kubectl to deploy a simple Kubernetes application:

    kubectl create deployment hello-node --image=gcr.io/hello-minikube-zero-install/hello-node

Creating required AMQ resources
authenticationservice.admin.enmasse.io/none-authservice created
addressspace.enmasse.io/workload-app created
address.enmasse.io/workload-app.queue-requests created
Waiting for AMQ AddressSpace to be ready
Waiting for AMQ Address to be ready
3scale.sh: API_URL=https://3scale-admin.apps.mnairn.l9z7.s1.devshift.org/admin/api
3scale.sh: create workload-app-api-account
3scale.sh: create workload-app-api-backend
3scale.sh: create workload-app-api-metric in backend=16
3scale.sh: create mapping_rule with backend_id=16 and metric_id=36
3scale.sh: create workload-app-api service
3scale.sh: error: xget: failed to find '//service/id/text()' in: '<?xml version="1.0" encoding="UTF-8"?><errors><error>System name has already been taken</error></errors>'
3scale.sh: create backend_usage with service_id= and backend_id=16
3scale.sh: create workload-app-api-plan in service=
3scale.sh: error: xget: failed to find '//plan/id/text()' in: '<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>Not Found</title>
    <link href="/assets/error.css" media="screen" rel="stylesheet" type="text/css" />
  </head>

  <body>
    <div id="wrapper">
      <div id="content">
        <h1 class='logo-3scale logo-3scale--svg logo-3scale--boxed' id="logo">
          3scale
        </h1>

        <h1>Not Found</h1>
        <p>Sorry. We can't find what you're looking for.</p>
      </div>

      <div id="footer">
        <ul>
          <li class="powered-by-3scale last">
            <a href="http://www.3scale.net" target="_blank">
              Powered by <span class="logo-3scale logo-3scale--svg logo-3scale--powered">3scale</span>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </body>
</html>'
3scale.sh: create workload-app-api-app in account=11 with plan_id=
3scale.sh: error: xget: failed to find '//application/user_key/text()' in: ''
Deploying the webapp with the following parameters:
AMQ_ADDRESS=amqps://messaging-eecbf30.redhat-rhmi-amq-online.svc
AMQ_QUEUE=/queue-requests
THREE_SCALE_URL=
deploymentconfig.apps.openshift.io/workload-web-app created
service/workload-web-app created
Waiting for pod to be ready
pod/workload-web-app-1-h5n5c condition met
route.route.openshift.io/metrics exposed

```

**Comment by b1zzu** [comment]
Here is the problem:
```
3scale.sh: create workload-app-api service
3scale.sh: error: xget: failed to find '//service/id/text()' in: '<?xml version="1.0" encoding="UTF-8"?><errors><error>System name has already been taken</error></errors>'
```
try to `make local/undeploy` wait for >10m and run `make local/deploy` again

**Comment by mikenairn** [comment]
After leaving it a while after running an undeloy, it seemed to work as expected:

```
$ oc expose svc/workload-web-app --path /metrics --name metrics
$ curl -s http://metrics-workload-web-app.apps.mnairn.l9z7.s1.devshift.org/metrics | grep ^three
three_scale_api_requests_success 1
```

Disrupting the pods also updates the metric as expected, 0 when requests fail.

/lgtm

**Comment by wei-lee** [comment]
@b1zzu if you can rebase the PR I will then approve it

**Pull request merged**